### ISSUE TITLE
sort function of typed list is very slow

### ISSUE THREAD

**Issue by Rinoahu** [numba/numba#5822]
Dear developers,

Why is the sort function of typed list almost 10 times slower than that of the reflected list?

```python
import numba as nb

@nb.njit
def mk_typed_list(n):
    x = nb.typed.List()

    for i in range(n):
        x.append(np.random.randint(0, 100))
    return x

X_t = mk_typed_list(10**7)
X = list(X_t)

%prun X_t.sort()
%prun X.sort()
```
Sorting X_t need 23s while sorting X only took 2.3s

**Comment by gmarkall** [comment]
For me, it took 21.861 seconds to sort the typed list and 0.034 seconds to sort the Python list.

Also converting the typed list to a Python list (`X = list(X_t)`) took about 9 seconds.

@esc Any thoughts on this?

**Comment by esc** [comment]
@Rinoahu  thanks for reporting this. I can reproduce this locally but I am not yet sure about the root cause. It may be related in some way to https://github.com/numba/numba/pull/5184. It's a bit tricky to benchmark this as  `sort` will sort in-place. The time taken to convert from `numba.typed.List` to Python `list` does look suspicious, but do bear in mind that every single integer needs to be boxed from a primitive/plain representation to a Python object. Also suspicious is that the `numba.typed.List` sort uses code from the reflected list sort implementation, which is here https://github.com/numba/numba/blob/master/numba/typed/listobject.py#L1224-L1263 -- so, I concur with the assessment, that "it shouldn't be this slow".

**Comment by Rinoahu** [comment]
> @Rinoahu thanks for reporting this. I can reproduce this locally but I am not yet sure about the root cause. It may be related in some way to #5184. It's a bit tricky to benchmark this as `sort` will sort in-place. The time taken to convert from `numba.typed.List` to Python `list` does look suspicious, but do bear in mind that every single integer needs to be boxed from a primitive/plain representation to a Python object. Also suspicious is that the `numba.typed.List` sort uses code from the reflected list sort implementation, which is here https://github.com/numba/numba/blob/master/numba/typed/listobject.py#L1224-L1263 -- so, I concur with the assessment, that "it shouldn't be this slow".

Actually, accessing the typed list is 10~100 times slower than the reflected list on my local laptop.

**Comment by esc** [comment]
@Rinoahu thank you for asking about this, and yes we are continuing to improve the `numba.typed.List` both in terms of features, usability and performance. I'd be interested to see some of your benchmarks beyond the sorting ("accessing the typed list is 10~100 times slower"), if you could share them? It would be useful as a starting point for further optimizations.

**Comment by Rinoahu** [comment]
> @Rinoahu thank you for asking about this, and yes we are continuing to improve the `numba.typed.List` both in terms of features, usability and performance. I'd be interested to see some of your benchmarks beyond the sorting ("accessing the typed list is 10~100 times slower"), if you could share them? It would be useful as a starting point for further optimizations.

Hi, sec, thank you for your reply. I wrote two simple functions that swap the element of a list.

```python
import numba as nb
import numpy as np

@nb.njit
def swap_reflect(X):
    n = len(X)
    x = []
    x.extend(X)
    for i in range(1, n):
        x[i-1], x[i] = x[i], x[i-1]

@nb.njit
def swap_typed(X):
    n = len(X)
    x = nb.typed.List()
    x.extend(X)

    for i in range(1, n):
        x[i-1], x[i] = x[i], x[i-1]

x = np.random.randn(10**6)

%timeit swap_reflect(x)
%timeit swap_typed(x)
```
I ran this code on google colab and the results are shown below:
```
The slowest run took 9.96 times longer than the fastest. This could mean that an intermediate result is being cached.
100 loops, best of 3: 13.2 ms per loop
1 loop, best of 3: 193 ms per loop```
```

**Comment by esc** [comment]
@Rinoahu thanks for getting back to me, I can reproduce this locally and see the following timings:

```
In [2]: %timeit swap_reflect(x)
   ...: %timeit swap_typed(x)
10.8 ms ± 588 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
135 ms ± 4.72 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by esc** [comment]
Note: I tried the following snippets to, to see if it would be related to `getitem` and `setitem`:

```
@nb.njit
def acc_reflect(X):
    acc = 0.0
    n = len(X)
    x = list(X)
    for i in range(n):
        acc += x[i]
    return acc


@nb.njit
def acc_typed(X):
    acc = 0.0
    n = len(X)
    x = nb.typed.List(X)

    for i in range(n):
        acc += x[i]
    return acc
```

And I get the following:

```
In [3]: %timeit acc_reflect(x)
   ...: %timeit acc_typed(x)
9.53 ms ± 81.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
51.3 ms ± 1.06 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

So, my current assumption is that both `getitem` and `setitem` seem to contribute equally to the slowness.

**Comment by Rinoahu** [comment]
@esc, yes, it is. Maybe this is why sorting a typed list is slow because sorting needs a lot of operations of getitem and setitem.

**Comment by esc** [comment]
@Rinoahu indeed, I think the next place to look is the `getitem` and `setitem` implementations.

**Comment by Rinoahu** [comment]
@esc Maybe other attributes of the typed list such as ```extend``` and ```append``` also need to be improved. These two functions run 3 times slower on a typed list than a reflect list.

**Comment by esc** [comment]
@Rinoahu indeed, both `append` and `extend` may use `getitem` and `setitem` under the hood, so they are likely related. The  implementations are significantly different. I tried the following patch:

```
diff --git i/numba/core/runtime/nrtdynmod.py w/numba/core/runtime/nrtdynmod.py
index a4a5462199..63aa52e556 100644
--- i/numba/core/runtime/nrtdynmod.py
+++ w/numba/core/runtime/nrtdynmod.py
@@ -102,7 +102,7 @@ def _define_nrt_decref(module, atomic_decr):

 # Set this to True to measure the overhead of atomic refcounts compared
 # to non-atomic.
-_disable_atomicity = 0
+_disable_atomicity = 1


 def _define_atomic_inc_dec(module, op, ordering):
```

And in my case it will cut the time for the typed-list in the reproducer in half. The next steps will be to take a closer look and possible revise how refcounting is handled.
